hyperspy/io.py

```diff
@@ -315,26 +291,26 @@ def assign_signal_subclass(dtype,
         dtype = 'real'
     else:
         raise ValueError('Data type "{}" not understood!'.format(dtype.name))
-    if record_by and record_by not in ["image", "spectrum"]:
-        raise ValueError("record_by must be one of: None, empty string, "
-                         "\"image\" or \"spectrum\"")
+    if not isinstance(signal_dimension, int) or signal_dimension < 0:
+        raise ValueError("signal_dimension must be a positive interger")
     signals = hyperspy.misc.utils.find_subclasses(hyperspy.signals, BaseSignal)
     d_matches = [s for s in signals.values() if dtype == s._dtype]
-    d_r_matches = [s for s in d_matches if record_by == s._record_by]
+    d_r_matches = [s for s in d_matches
```

Comment:
A very minor nitpick: I chose the very short names because of the (then) long list of tested attributes:
- d: `dtype`
- r: `record_by` (now `signal_dimension`)
- o: `signal_origin` (now gone)
- t: `signal_type`

After `signal_origin` is gone, maybe we should go back to more verbose names like `dtype_matches`, `dtype_dim_matches` and `dtype_dim_type_matches`, which may be easier to read for new people.